devicetypes/zenwithin/zen-thermostat.src/zen-thermostat.groovy

```diff
@@ -3,546 +3,808 @@
         }
-        valueTile("heatingSetpoint", "device.heatingSetpoint", inactiveLabel: false) {
-			state "heat", label:'${currentValue}° heat', unit:"F", backgroundColor:"#ffffff"
-		}
-        valueTile("coolingSetpoint", "device.coolingSetpoint", inactiveLabel: false) {
-			state "cool", label:'${currentValue}° cool', unit:"F", backgroundColor:"#ffffff"
-		}
-        standardTile("thermostatOperatingState", "device.thermostatOperatingState", inactiveLabel: false) {
-            state "heating", backgroundColor:"#ff6e7e"
-            state "cooling", backgroundColor:"#90d0e8"
-            state "fan only", backgroundColor:"#e8e3d8"
-		}
-        standardTile("setpointUp", "device.thermostatSetpoint", decoration: "flat") {
-            state "setpointUp", action:"setpointUp", icon:"st.thermostat.thermostat-up"
+        standardTile("fanMode", "device.thermostatFanMode", width:2, height:2, inactiveLabel: false, decoration: "flat") {
+            state "speed", label:'${currentValue}', action:"switchFanMode", nextState:"updating", icon: "st.thermostat.fan-on"
+            state "auto", action:"switchFanMode", nextState:"updating", icon: "st.thermostat.fan-auto"
+            state "on", action:"switchFanMode", nextState:"updating", icon: "st.thermostat.fan-on"
+            state "updating", label: "Updating...", nextState:"updating", backgroundColor:"#ffffff"
         }
-        
-        standardTile("setpointDown", "device.thermostatSetpoint", decoration: "flat") {
-            state "setpointDown", action:"setpointDown", icon:"st.thermostat.thermostat-down"
+        standardTile("thermostatOperatingState", "device.thermostatOperatingState", width: 2, height:2, decoration: "flat") {
+            state "thermostatOperatingState", label:'${currentValue}', backgroundColor:"#ffffff"
         }
-
-        standardTile("refresh", "device.temperature", decoration: "flat") {
+        standardTile("refresh", "device.thermostatMode", width:2, height:2, inactiveLabel: false, decoration: "flat") {
             state "default", action:"refresh.refresh", icon:"st.secondary.refresh"
         }
-        
-        standardTile("configure", "device.configure", decoration: "flat") {
-            state "configure", label:'', action:"configuration.configure", icon:"st.secondary.configure"
+        main "temperature"
+        details(["temperature", "mode", "fanMode", "thermostatOperatingState", "refresh"])
+    }
+    preferences {
+        section {
+            input("systemModes", "enum",
+                title: "Thermostat configured modes\nSelect the modes the thermostat has been configured for, as displayed on the thermostat",
+                description: "off, heat, cool", defaultValue: "1", required: true, multiple: false,
+                options:["1":"off, heat, cool", "2":"off, auto, heat, cool", "3":"off, emergency heat, heat, cool"]
+            )
         }
+    }
+}
+
+// Globals
+private getTHERMOSTAT_CLUSTER()                      { 0x0201 }
```

Comment:
It looks like you decided against using any of these since I don't really see them called anywhere.  Consider removing

Reply:
Nevermind, you're using them in pollDevice().  But why not anywhere else?

Reply:
Two reasons, when using e.g. THERMOSTAT_CLUSTER it always result in a call to this method, the compilers isn't smart enough to realize that it can substitute with the const returned value, and then it has to be converted to a string as that is what the comparisons are made agains, although I could possible have converted the incoming value to hex too, event so there's a lot of comparisons in the parse method that would generate a lot of calls to theses getters.